### PR TITLE
resize before loading previews

### LINKED ISSUE #922: Offer warning before generating preview "quick-add steps" on very large images
Uploading a big image (actually like any image from a smartphone camera 😂 📸 🤳 ) stalls out the sequencer because of the preview images. A few ideas:

1. we could not generate previews for the small steps if the image size is over a certain threshold, instead showing a message `Generating previews for images this large may take a while. _Do it anyways_?` so it can be overridden...
2. do we generate the preview resize once, then pass it to each preview step, or do we generate it multiple times from the original image? 
3. make resizing images faster with optimization of the resize module?

Just thinking how we could smooth this out a bit!

### PR DESCRIPTION
**Pull request by aashna27**
Fixes #922 & #1122 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `npm test`
* [ ] code is in uniquely-named feature branch and has no merge conflicts
* [ ] PR is descriptively titled
* [ ] ask `@publiclab/is-reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

Thanks!

### CODE REVIEW

**Comment by aashna27** [comment]
Now the webpage doesnt fail and gives the preview even for very large image but it does take some time
Crashing-Big

**Comment by Divy123** [comment]
@aashna27 is it based on what we discussed?

**Comment by aashna27** [comment]
> @aashna27 is it based on what we discussed?
Yeah sorry I forgot to mention you I made an earlier one today and did mention you there!! :speak_no_evil: 
Yeah so really thankful to @Divy123, he was patient and helped me with what I was stuck with!! :hugs:

**Comment by Divy123** [comment]
@aashna27 thanks for your appreciation.
Though I believe we are a team and it feels great to see this working. I think you have done a fantastic job and removed a big restriction to the usage of IS.

So, cheers for this.

**Comment by aashna27** [comment]
> @aashna27 thanks for your appreciation.
> Though I believe we are a team and it feels great to see this working. I think you have done a fantastic job and removed a big restriction to the usage of IS.
> 
> So, cheers for this.

Hehe Thanks @Divy123  for motivating and also please guide me for your pr, I will look for sources!!

**Comment by Divy123** [comment]
Yeah sure. I will leave some comments today and clearly state what I am
currently working on

Thanks a lot once again.

**Comment by jywarren** [comment]
This is awesome teamwork!

**Comment by jywarren** [comment]
Is this good to merge? Very exciting!!

**Comment by aashna27** [comment]
Yeah I believe that it is good to merge!

File: examples/lib/insertPreview.js [hunk 1]
```diff
@@ -44,9 +47,17 @@ function updatePreviews(src, selector) {
     }
   };
 
-  Object.keys(previewSequencerSteps).forEach(function (step, index) {
-    generatePreview(step, Object.values(previewSequencerSteps)[index], src, selector);
+  var sequencer = ImageSequencer();
+
+  sequencer.loadImage(src, function(){
+    this.addSteps('resize', {['resize']: '30%'});
```
Comment (jywarren) [inline examples/lib/insertPreview.js]: Should we limit to a fixed size? Like 100x100? Because really large images at 30% are still quite large!

Comment (aashna27) [reply examples/lib/insertPreview.js]: but idt we have that option in resize.

**Comment by jywarren** [comment]
Can we calculate it using image width?

**Comment by aashna27** [comment]
Hey @jywarren I thought about this and I believe that for this I need to change the input values for resize! And take it as 2 dimensions width and height, instead of the currently accepting percentage value. Do you think there is any other way?

**Comment by jywarren** [comment]
Hmm. I guess I thought based on image dimensions we could calculate the
percentage. But the resize module should perhaps accept absolute dimensions
anyways. Maybe we can set a units parameter? As either pixels or
percentage?

But what do you think of merging this first then doing that as a follow-up?

**Comment by aashna27** [comment]
Yeah actually that was my point but the i thought that for that we will
need our inputs to be dynamic, ie if user selects percentage then just ask
for perc value and if absolute then ask height and width. And this would
require great changes to the currently accepting input js we have!

But i am thinking maybe give the src to dimensions a try. And if that
doesn't work out we can merge this and I will do the absolute dimensions in
a follow-up.

Relying on this
https://stackoverflow.com/questions/17774928/js-get-image-width-and-height-from-the-base64-code
for dimensions from img src.

**Comment by aashna27** [comment]
@jywarren I have made the changes and tested it on all types of image sizes.
big-prev

cyan-prev

tulip-prev

**Comment by jywarren** [comment]
Awesome!!!!

**Comment by jywarren** [comment]
Super. We can merge to gh-pages soon!

**Comment by aashna27** [comment]
Yeah many new features

**Pull request merged**